File: 백준/Platinum/5419. 북서풍/북서풍.java
import java.io.*;
import java.util.ArrayList;

import static java.util.Comparator.comparingInt;

public class Main {
    static int[] buildTree(int n) {
        int capacity = 1;
        while (capacity < n)
            capacity *= 2;
        int[] ret = new int[capacity << 1];
        ret[0] = capacity;
        return ret;
    }

    static void update(int[] tree, int y) {
        int node = tree[0] + y;
        while (node > 0) {
            tree[node]++;
            node >>= 1;
        }
    }

    static int query(int[] tree, int qs, int qe) {
        int start = tree[0] + qs;
        int end = tree[0] + qe;
        int ret = 0;
        while (start < end) {
            if ((start & 1) == 1)
                ret += tree[start++];
            if ((end & 1) == 0)
                ret += tree[end--];
            start >>= 1;
            end >>= 1;
        }
        if (start == end)
            ret += tree[start];
        return ret;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int testCase = Integer.parseInt(br.readLine());

        while (testCase-- > 0) {
            int n = Integer.parseInt(br.readLine());
            ArrayList<int[]> islands = new ArrayList<>();

            for (int i = 0; i < n; i++) {
                String[] island = br.readLine().split(" ");
                islands.add(new int[] { Integer.parseInt(island[0]), Integer.parseInt(island[1]) });
            }

            islands.sort(comparingInt((int[] a) -> a[1]).reversed());

            int count = 0;
            for (int i = n - 1; i > 0; i--) {
                if (islands.get(i)[1] == islands.get(i - 1)[1]) {
                    islands.get(i)[1] = count;
                } else {
                    islands.get(i)[1] = count++;
                }
            }
            islands.get(0)[1] = count;
            islands.sort(comparingInt((int[] a) -> a[0]));

            int[] tree = buildTree(count + 1);
            long answer = 0L;
            for (int[] island : islands) {
                answer += query(tree, island[1], count);
                update(tree, island[1]);
            }

            bw.write(Long.toString(answer));
            bw.newLine();
        }

        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Platinum/5419. 북서풍/북서풍.kt
import java.util.*

fun main() {
    val br = System.`in`.bufferedReader()
    val bw = System.out.bufferedWriter()
    val testCase = br.readLine().toInt()

    repeat(testCase) {
        val n = br.readLine().toInt()
        val islands = MutableList(n) {
            val line = StringTokenizer(br.readLine())
            Island(line.nextToken().toInt(), line.nextToken().toInt())
        }

        islands.sortByDescending { it.y }
        var count = 0
        for (i in n - 1 downTo 1) {
            if (islands[i].y == islands[i - 1].y) {
                islands[i].y = count
            } else {
                islands[i].y = count++
            }
        }
        islands[0].y = count
        islands.sortBy { it.x }

        var answer = 0L
        val nw = Northwester(count + 1)
        for (e in islands) {
            answer += nw.query(e.y, count)
            nw.update(e.y)
        }
        bw.write(answer.toString())
        bw.newLine()
    }
    bw.flush()
    bw.close()
    br.close()
}

data class Island(var x: Int, var y: Int)

class Northwester(n: Int) {
    val tree: IntArray
    var capacity = 1

    init {
        while (capacity < n) capacity *= 2
        tree = IntArray(capacity shl 1)
    }

    fun update(y: Int) {
        var node = capacity + y
        while (node > 0) {
            tree[node]++
            node = node ushr 1
        }
    }

    fun query(qs: Int, qe: Int): Int {
        var start = capacity + qs
        var end = capacity + qe
        var ret = 0
        while (start < end) {
            if (start and 1 == 1) ret += tree[start++]
            if (end and 1 == 0) ret += tree[end--]
            start = start ushr 1
            end = end ushr 1
        }
        if (start == end) ret += tree[start]
        return ret
    }
}